### PR TITLE
Bluetooth: Audio: broadcast assistant sync to PA

### PR DESCRIPTION
**Pull request by pin-zephyr**
broadcast_audio_assistant application sync to PA and obtain the correct subgroup information used by BASS.

### CODE REVIEW

File: samples/bluetooth/broadcast_audio_assistant/src/main.c [hunk 1]
```diff
@@ -121,6 +126,84 @@ static bool device_found(struct bt_data *data, void *user_data)
...
+	/* Compare BASE and copy if different */
+	if (base_size != received_base_size || memcmp(base, received_base, base_size) != 0) {
+		(void)memcpy(received_base, base, base_size);
+		received_base_size = base_size;
+	}
+
+	/* Stop parsing */
+	k_sem_give(&sem_received_base_subgroups);
+	return false;
+}
```
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_assistant/src/main.c:150]: The name of this function does not match the behavior. It doesn't decode anything, but seemingly just stored the BASE for later use

File: samples/bluetooth/broadcast_audio_assistant/src/main.c [hunk 2]
```diff
@@ -121,6 +126,84 @@ static bool device_found(struct bt_data *data, void *user_data)
...
+
+static inline bool add_pa_sync_base_subgroup_bis_cb(const struct bt_bap_base_subgroup_bis *bis,
```
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_assistant/src/main.c:152]:
```suggestion
static bool add_pa_sync_base_subgroup_bis_cb(const struct bt_bap_base_subgroup_bis *bis,
```

Generally there are rarely any reason to use `inline` for static function.

File: samples/bluetooth/broadcast_audio_assistant/src/main.c [hunk 3]
```diff
@@ -121,6 +126,84 @@ static bool device_found(struct bt_data *data, void *user_data)
...
+}
+
+static void pa_recv(struct bt_le_per_adv_sync *sync,
+			 const struct bt_le_per_adv_sync_recv_info *info,
+			 struct net_buf_simple *buf)
+{
+	bt_data_parse(buf, pa_decode_base, NULL);
+}
+
+static inline bool add_pa_sync_base_subgroup_cb(const struct bt_bap_base_subgroup *subgroup,
```
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_assistant/src/main.c:169]:
```suggestion
static bool add_pa_sync_base_subgroup_cb(const struct bt_bap_base_subgroup *subgroup,
```

Generally there are rarely any reason to use `inline` for static function.

File: samples/bluetooth/broadcast_audio_assistant/src/main.c [hunk 4]
```diff
@@ -121,6 +126,84 @@ static bool device_found(struct bt_data *data, void *user_data)
...
+	struct bt_bap_broadcast_assistant_add_src_param *param = user_data;
+	struct bt_bap_bass_subgroup *subgroup_param;
+	uint8_t *data;
+	int ret;
+
+	if (param->num_subgroups == CONFIG_BT_BAP_BASS_MAX_SUBGROUPS) {
+		printk("Cannot fit all subgroups param with size %d",
+			CONFIG_BT_BAP_BASS_MAX_SUBGROUPS);
+		return true; /* return true to avoid returning -ECANCELED as this is OK */
+	}
```
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_assistant/src/main.c:181]: Seems like it makes more sense to do these checks before we store the BASE, to avoid unnecessary copies

File: samples/bluetooth/broadcast_audio_assistant/src/main.c [hunk 5]
```diff
@@ -121,6 +126,84 @@ static bool device_found(struct bt_data *data, void *user_data)
...
+		return true;
+	}
+
+	base_size = data->data_len - BT_UUID_SIZE_16; /* the BASE comes after the UUID */
+
+	/* Compare BASE and copy if different */
+	if (base_size != received_base_size || memcmp(base, received_base, base_size) != 0) {
+		(void)memcpy(received_base, base, base_size);
+		received_base_size = base_size;
+	}
```
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_assistant/src/main.c:145]: Do we need to guard this with a mutex or something, so that we avoid overwriting the BASS while we call `bt_bap_base_foreach_subgroup`?

File: samples/bluetooth/broadcast_audio_assistant/src/main.c [hunk 6]
```diff
@@ -121,6 +126,84 @@ static bool device_found(struct bt_data *data, void *user_data)
...
+	ret = bt_bap_base_get_subgroup_codec_meta(subgroup, &data);
+	if (ret < 0) {
+		return false;
+	}
+
+	subgroup_param = &param->subgroups[param->num_subgroups];
+
+	if (ret > ARRAY_SIZE(subgroup_param->metadata)) {
+		printk("Cannot fit %d octets into subgroup param with size %zu", ret,
+			   ARRAY_SIZE(subgroup_param->metadata));
```
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_assistant/src/main.c:192]:
```suggestion
		printk("Cannot fit %d octets into subgroup param with size %zu", ret,
		       ARRAY_SIZE(subgroup_param->metadata));
```

File: samples/bluetooth/broadcast_audio_assistant/src/main.c [hunk 7]
```diff
@@ -121,6 +126,84 @@ static bool device_found(struct bt_data *data, void *user_data)
...
+						void *user_data)
+{
+	struct bt_bap_broadcast_assistant_add_src_param *param = user_data;
+	struct bt_bap_bass_subgroup *subgroup_param;
+	uint8_t *data;
+	int ret;
+
+	if (param->num_subgroups == CONFIG_BT_BAP_BASS_MAX_SUBGROUPS) {
+		printk("Cannot fit all subgroups param with size %d",
+			CONFIG_BT_BAP_BASS_MAX_SUBGROUPS);
```
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_assistant/src/main.c:179]:
```suggestion
		printk("Cannot fit all subgroups param with size %d",
		       CONFIG_BT_BAP_BASS_MAX_SUBGROUPS);
```

File: samples/bluetooth/broadcast_audio_assistant/src/main.c [hunk 8]
```diff
@@ -143,6 +226,19 @@ static bool is_substring(const char *substr, const char *str)
...
 
+static int pa_sync_create(void)
+{
+	struct bt_le_per_adv_sync_param create_params = {0};
+
+	bt_addr_le_copy(&create_params.addr, &selected_addr);
+	create_params.options = BT_LE_PER_ADV_SYNC_OPT_FILTER_DUPLICATE;
+	create_params.sid = selected_sid;
+	create_params.skip = PA_SYNC_SKIP;
+	create_params.timeout = selected_pa_interval;
```
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_assistant/src/main.c:237]: If you set the timeout to `selected_pa_interval`, then if it missing a single PA report, then it would fail to sync. 

Furthermore, `selected_pa_interval` are in units of of 1.25ms, while `timeout` is in `10ms`.

File: samples/bluetooth/broadcast_audio_assistant/src/main.c [hunk 9]
```diff
@@ -199,6 +295,9 @@ static void scan_recv_cb(const struct bt_le_scan_recv_info *info,
 			bt_addr_le_copy(&selected_addr, info->addr);
 
 			k_sem_give(&sem_source_discovered);
+
+			printk("Attempting to PA sync to the broadcaster with id 0x%06X\n",
+				selected_broadcast_id);
```
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_assistant/src/main.c:300]:
```suggestion
			printk("Attempting to PA sync to the broadcaster with id 0x%06X\n",
			       selected_broadcast_id);
```

File: samples/bluetooth/broadcast_audio_assistant/src/main.c [hunk 10]
```diff
@@ -369,11 +468,33 @@ static void bap_broadcast_assistant_add_src_cb(struct bt_conn *conn, int err)
...
 }
 
+static void pa_sync_synced_cb(struct bt_le_per_adv_sync *sync,
+			      struct bt_le_per_adv_sync_synced_info *info)
+{
+	if (sync == pa_sync ||
+	    (info != NULL && bt_addr_le_eq(info->addr, &selected_addr) &&
+	     info->sid == selected_sid)) {
+		printk("PA sync %p synced for broadcast sink with broadcast ID 0x%06X\n", sync,
+		selected_broadcast_id);
```
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_assistant/src/main.c:478]:
```suggestion
		printk("PA sync %p synced for broadcast sink with broadcast ID 0x%06X\n", sync,
		       selected_broadcast_id);
```

File: samples/bluetooth/broadcast_audio_assistant/src/main.c [hunk 11]
```diff
@@ -369,11 +468,33 @@ static void bap_broadcast_assistant_add_src_cb(struct bt_conn *conn, int err)
 	}
 }
 
+static void pa_sync_synced_cb(struct bt_le_per_adv_sync *sync,
+			      struct bt_le_per_adv_sync_synced_info *info)
+{
+	if (sync == pa_sync ||
+	    (info != NULL && bt_addr_le_eq(info->addr, &selected_addr) &&
+	     info->sid == selected_sid)) {
```
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_assistant/src/main.c:476]:
```suggestion
	if (sync == pa_sync) {
```

Just checking the sync object should be fine

File: samples/bluetooth/broadcast_audio_assistant/src/main.c [hunk 12]
```diff
@@ -369,11 +468,33 @@ static void bap_broadcast_assistant_add_src_cb(struct bt_conn *conn, int err)
...
+		printk("PA sync %p synced for broadcast sink with broadcast ID 0x%06X\n", sync,
+		selected_broadcast_id);
+
+		if (pa_sync == NULL) {
+			pa_sync = sync;
+		}
```
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_assistant/src/main.c:482]: This should never happen
```suggestion
```

File: samples/bluetooth/broadcast_audio_assistant/src/main.c [hunk 13]
```diff
@@ -467,31 +590,38 @@ int main(void)
...
-		 * Also, if the source contains more than one stream (e.g. stereo),
-		 * some sinks have been observed to have issues. In this case,
-		 * set only one bit in bis_sync, e.g. subgroup.bis_sync = BIT(1).
-		 *
-		 * When PA sync and BASE is parsed (see note in the scan_recv_cb function),
-		 * the available bits can be used for proper selection.
-		 */
-		subgroup.bis_sync = BT_BAP_BIS_SYNC_NO_PREF;
+		printk("Scan stopped, attempting to PA sync to the broadcaster with id 0x%06X\n",
+			selected_broadcast_id);
```
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_assistant/src/main.c:594]:
```suggestion
		printk("Scan stopped, attempting to PA sync to the broadcaster with id 0x%06X\n",
		       selected_broadcast_id);
```

File: samples/bluetooth/broadcast_audio_assistant/src/main.c [hunk 14]
```diff
@@ -467,31 +590,38 @@ int main(void)
...
+			printk("Could not create Broadcast PA sync: %d, resetting\n", err);
+			return -ETIMEDOUT;
+		}
+
+		printk("Waiting for PA synced\n");
+		err = k_sem_take(&sem_pa_synced, K_FOREVER);
+		if (err != 0) {
+			printk("sem_pa_synced timed out, resetting\n");
+			return -ETIMEDOUT;
+		}
```
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_assistant/src/main.c:606]: Does it make sense to return `-ETIMEDOUT` when using `K_FOREVER`? Those seem mutually exclusive

File: samples/bluetooth/broadcast_audio_assistant/src/main.c [hunk 15]
```diff
@@ -467,31 +590,38 @@ int main(void)
...
-		param.num_subgroups = 1;
-		param.subgroups = &subgroup;
+		/* Wait to receive subgroups */
+		err = k_sem_take(&sem_received_base_subgroups, K_FOREVER);
+		if (err != 0) {
+			printk("Failed to take sem_received_base_subgroups (err %d)\n", err);
+		}
+
+		err = bt_bap_base_foreach_subgroup((const struct bt_bap_base *)received_base,
+						   add_pa_sync_base_subgroup_cb, &param);
```
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_assistant/src/main.c:621]: If we have not receive a BASE, then we should not yet start parsing it

File: samples/bluetooth/broadcast_audio_assistant/src/main.c [hunk 16]
```diff
@@ -467,31 +590,38 @@ int main(void)
...
+		if (err < 0) {
+			printk("Could not add BASE to params %d\n", err);
+		}
```
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_assistant/src/main.c:624]: Should we stop from continuing here?

File: include/zephyr/bluetooth/audio/bap.h [hunk 17]
```diff
@@ -2239,7 +2239,7 @@ struct bt_bap_broadcast_assistant_mod_src_param {
 	uint8_t num_subgroups;
 
 	/** Pointer to array of subgroups */
-	struct bt_bap_bass_subgroup *subgroups;
+	struct bt_bap_bass_subgroup subgroups[CONFIG_BT_BAP_BASS_MAX_SUBGROUPS];
```
Comment (Thalley) [inline include/zephyr/bluetooth/audio/bap.h:2242]: The changes in this file significantly increase the size of the parameters, and can cause some stack size issues. 

Why is this change required?

**Review by Thalley — CHANGES_REQUESTED** [review]
It is OK to make an API change and a sample change in a single PR, but they should be split into multiple commits. Similar the changes to the bsim tests should also be a separate commit (if you intend to keep the changes in the header file).

File: samples/bluetooth/broadcast_audio_assistant/prj.conf [hunk 18]
```diff
@@ -11,4 +11,8 @@ CONFIG_BT_CTLR_SCAN_DATA_LEN_MAX=191
 CONFIG_BT_TINYCRYPT_ECC=y
 
 CONFIG_BT_EXT_ADV=y
+CONFIG_BT_ISO_MAX_CHAN=2
```
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_assistant/prj.conf:14]: The broadcast assistant don't need ISO at all - Why are we setting ISO chan to 2?

File: samples/bluetooth/broadcast_audio_assistant/src/main.c [hunk 19]
```diff
@@ -17,7 +17,8 @@
 #include <zephyr/sys/byteorder.h>
 
 #define NAME_LEN 30
-
+#define PA_SYNC_SKIP         5
+#define SYNC_RETRY_COUNT     6 /* similar to retries for connections */
```
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_assistant/src/main.c:21]: In https://github.com/zephyrproject-rtos/zephyr/pull/70110 I modified these values for most other samples and tests. 

For the Zephyr controller it only skips up until there are 6 events left, which effectively means that `SYNC_RETRY_COUNT` of 6 or lower does not do anything

Consider applying similar changes from my PR here

File: samples/bluetooth/broadcast_audio_assistant/src/main.c [hunk 20]
```diff
@@ -121,6 +129,86 @@ static bool device_found(struct bt_data *data, void *user_data)
...
+
+static void pa_recv(struct bt_le_per_adv_sync *sync,
+			 const struct bt_le_per_adv_sync_recv_info *info,
+			 struct net_buf_simple *buf)
+{
+	bt_data_parse(buf, base_store, NULL);
+}
+
+static bool add_pa_sync_base_subgroup_bis_cb(const struct bt_bap_base_subgroup_bis *bis,
+						    void *user_data)
```
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_assistant/src/main.c:171]:
```suggestion
static bool add_pa_sync_base_subgroup_bis_cb(const struct bt_bap_base_subgroup_bis *bis,
					     void *user_data)
```

File: samples/bluetooth/broadcast_audio_assistant/src/main.c [hunk 21]
```diff
@@ -121,6 +129,86 @@ static bool device_found(struct bt_data *data, void *user_data)
...
+{
+	struct bt_bap_bass_subgroup *subgroup_param = user_data;
+
+	subgroup_param->bis_sync |= BIT(bis->index);
+
+	return true;
+}
+
+static bool add_pa_sync_base_subgroup_cb(const struct bt_bap_base_subgroup *subgroup,
+						void *user_data)
```
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_assistant/src/main.c:181]:
```suggestion
static bool add_pa_sync_base_subgroup_cb(const struct bt_bap_base_subgroup *subgroup,
					 void *user_data)
```

File: samples/bluetooth/broadcast_audio_assistant/src/main.c [hunk 22]
```diff
@@ -467,31 +612,42 @@ int main(void)
...
-		param.subgroups = &subgroup;
+		/* Wait to receive subgroups */
+		err = k_sem_take(&sem_received_base_subgroups, K_FOREVER);
+		if (err != 0) {
+			printk("Failed to take sem_received_base_subgroups (err %d)\n", err);
+			return err;
+		}
+
+		err = bt_bap_base_foreach_subgroup((const struct bt_bap_base *)received_base,
+						   add_pa_sync_base_subgroup_cb, &param);
```
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_assistant/src/main.c:646]: `received_base` may change while we are parsing it. Consider guarding this call and `store_base` with a mutex or similar.

**Review by Thalley — CHANGES_REQUESTED** [review]
Looks good; a few minor requests

File: samples/bluetooth/broadcast_audio_assistant/prj.conf [hunk 23]
```diff
@@ -11,4 +11,7 @@ CONFIG_BT_CTLR_SCAN_DATA_LEN_MAX=191
 CONFIG_BT_TINYCRYPT_ECC=y
 
 CONFIG_BT_EXT_ADV=y
+CONFIG_BT_BAP_BASS_MAX_SUBGROUPS=2
 CONFIG_BT_BAP_BROADCAST_ASSISTANT=y
+CONFIG_BT_AUDIO_CODEC_CFG_MAX_METADATA_SIZE=255
+CONFIG_SELECT_SOURCE_NAME="Broadcast Audio Source"
```
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_assistant/prj.conf:17]: It doesn't make sense to set `CONFIG_SELECT_SOURCE_NAME` to the non-default value in this file IMO. 

Suggest to remove this and keep the default to "" so it connects to the first again (typical behavior in our samples)

**Review by Thalley — DISMISSED** [review]

File: samples/bluetooth/broadcast_audio_assistant/prj.conf [hunk 24]
```diff
@@ -11,4 +11,7 @@ CONFIG_BT_CTLR_SCAN_DATA_LEN_MAX=191
 CONFIG_BT_TINYCRYPT_ECC=y
 
 CONFIG_BT_EXT_ADV=y
+CONFIG_BT_BAP_BASS_MAX_SUBGROUPS=2
 CONFIG_BT_BAP_BROADCAST_ASSISTANT=y
+CONFIG_BT_AUDIO_CODEC_CFG_MAX_METADATA_SIZE=255
+CONFIG_SELECT_SOURCE_NAME=""
```
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_assistant/prj.conf:17]:
```suggestion
```

This is the default value and can (should) be removed from here

**Review by Thalley — CHANGES_REQUESTED** [review]

**Review by Thalley — APPROVED** [review]

**Review by larsgk — APPROVED** [review]
